src/fifelse.c

```diff
@@ -201,105 +201,104 @@ SEXP fifelseR(SEXP l, SEXP a, SEXP b, SEXP na) {
   return ans;
 }
 
-SEXP fcaseR(SEXP na, SEXP rho, SEXP args) {
-  const int narg=length(args);
+SEXP fcaseR(SEXP rho, SEXP args) {
+  const int narg=length(args); // `default` will take the last two positions
   if (narg % 2) {
     error(_("Received %d inputs; please supply an even number of arguments in ..., "
             "consisting of logical condition, resulting value pairs (in that order). "
-            "Note that the default argument must be named explicitly, e.g., default=0"), narg);
+            "Note that the default argument must be named explicitly, e.g., default=0"), narg - 2);
   }
-  if (narg==0) return R_NilValue;
-
-  SEXP cons0 = PROTECT(eval(SEXPPTR_RO(args)[0], rho));
-  SEXP value0 = PROTECT(eval(SEXPPTR_RO(args)[1], rho)); // value0 will be compared to from loop so leave it protected throughout
-  SEXPTYPE type0 = TYPEOF(value0);
-  int64_t len0=xlength(cons0), len2=len0;
-  if (isS4(value0) && !INHERITS(value0, char_nanotime)) {
-    error(_("S4 class objects (except nanotime) are not supported. Please see https://github.com/Rdatatable/data.table/issues/4131."));
-    // otherwise 'invalid type/length (S4/1) in vector allocation' from test 2132.3
-  }
-  SEXP ans = PROTECT(allocVector(type0, len0));
-  SEXP tracker = PROTECT(allocVector(INTSXP, len0));
-  int *restrict p = INTEGER(tracker);
-  copyMostAttrib(value0, ans);
-
-  bool nonna=!isNull(na);
-  if (nonna) {
-    if (xlength(na) != 1) {
-      error(_("Length of 'default' must be 1."));
-    }
-    SEXPTYPE tn = TYPEOF(na);
-    if (tn==LGLSXP && LOGICAL(na)[0]==NA_LOGICAL) {
-      nonna = false;
-    } else {
-      if (tn != type0) {
-        error(_("Resulting value is of type %s but 'default' is of type %s. "
-                "Please make sure that both arguments have the same type."), type2char(type0), type2char(tn));
-      }
-      if (!R_compute_identical(PROTECT(getAttrib(value0,R_ClassSymbol)),  PROTECT(getAttrib(na,R_ClassSymbol)), 0)) {
-        error(_("Resulting value has different class than 'default'. "
-                "Please make sure that both arguments have the same class."));
-      }
-      UNPROTECT(2);
-      if (isFactor(value0)) {
-        if (!R_compute_identical(PROTECT(getAttrib(value0,R_LevelsSymbol)),  PROTECT(getAttrib(na,R_LevelsSymbol)), 0)) {
-          error(_("Resulting value and 'default' are both type factor but their levels are different."));
-        }
-        UNPROTECT(2);
-      }
-    }
-  }
-
+  int nprotect = 0, l = 0;
+  int64_t len0=0, len1=0, len2=0, idx=0;
+  SEXP ans = R_NilValue, value0 = R_NilValue, tracker = R_NilValue, whens = R_NilValue, thens = R_NilValue;
+  PROTECT_INDEX Iwhens, Ithens;
+  PROTECT_WITH_INDEX(whens, &Iwhens); nprotect++;
```

Comment:
I don't understand why `PROTECT_WITH_INDEX` is being used. Could a comment be added explaining what these variables are and why this protection method was chosen?

Reply:
The basic idea is to re-use memory as we go along, I tried naming these `when`/`then` to evoke `CASE WHEN A THEN B WHEN C THEN D ...`. So `PROTECT_INDEX` is to make the memory for `A`, `C`, `E`, ... is in the same place, same for `B`, `D`, `F`, ...

(AIUI)

I am not sure the benefits/tradeoffs here. The first commit in this branch was much simpler https://github.com/Rdatatable/data.table/pull/4264/commits/3334d0c3a137270d83c60108353c30a6b6e2e4d0

Reply:
I am still noob at `PROTECT_WITH_INDEX()` so I turned to LLM for some explanations...

https://g.co/gemini/share/85cdc98bb748
https://chatgpt.com/share/a4792977-f24a-47da-95c0-da48a1ad7c4c
Claude (permalinks not supported?) ![Claude Pt 1](https://github.com/Rdatatable/data.table/assets/7606389/02f1cddc-a7b2-447f-9e6b-91401222bdd6) 
![Claude Pt 2](https://github.com/Rdatatable/data.table/assets/7606389/84ce9e55-868c-47ed-be2f-5ae0ff2058b2)

The main themes are pretty consistent -- the alternative involves doing `UNPROTECT(2)` and then `PROTECT()` twice again in each iteration, hence potentially lots of RAM churn that we avoid with `PROTECT_WITH_INDEX()`.